### ISSUE TITLE
FFmpeg 5.0: internal audio player's threads hang

### ISSUE THREAD

**Issue by vedgy** [goldendict/goldendict#1489]
I have built latest GoldenDict master against FFmpeg 5.0 on Manjaro GNU/Linux and selected *FFmpeg+libao* internal player.

Most sounds fail to be pronounced. Furthermore, each failed pronunciation attempt increases GoldenDict's CPU usage. Finally, GoldenDict continues to hang and utilize the CPU cores when I try to exit it.

If this is not a local issue on my system, then perhaps we should prevent building GoldenDict against FFmpeg 5.0 until someone properly implements support for this library version.

**Comment by nonwill** [comment]
Here may be the issue:
https://github.com/goldendict/goldendict/blob/f810c6bd724e61977b4e94ca2d8abfa5bd766379/ffmpegaudio.cc#L146

If the pStream is at end, GoldenDict will keep calling readAudioData.  To end the reading(with no data), just add two lines of codes before the return :
![ffmpeg_loop_issue](https://user-images.githubusercontent.com/46510529/169740407-b1086c81-e434-40ac-ae57-4277fe476902.png)

**Comment by vedgy** [comment]
@nonwill, since you have done the bulk of the work already, could you please create a pull request to get the credit for it?

**Comment by nonwill** [comment]
No, I don't need the credit for it. It's just a suggestion, and in my opinion it's the easiest solution to fix the issue. 
Exactly, I'm not sure if it is a problem with the latest version of ffmpeg or not, as previous versions may not have had the problem, maybe it should be addressed in ffmpeg.
And in the latest version of GD++, the improvements and support for ffmpeg audioplayer have been deprecated.  Both Fmod and Bass audioplaye plugins have performed very consistently, with a small memory footprint and sound particularly good.

**Comment by vedgy** [comment]
Neither FMOD nor BASS has a license compatible with GoldenDict's GPLv3, so they cannot be used legally. What do you think about the Qt Multimedia backend?

**Comment by nonwill** [comment]
FMOD and BASS can be freely used in non-commercial scenarios.
Personally, I don't think too highly of the Qt Multimedia backend for that, at least on Window, it may require the user to install some decoding libraries themselves to run it, which adds complexity and can cause some compatibility issues. And on macOS playing a buffer does not work, you must first cache the audio content to a file and then play the file.

**Comment by vedgy** [comment]
> FMOD and BASS can be freely used in non-commercial scenarios.

But they are still incompatible with the GPL. So distributing GoldenDict that depends on any of these two libraries is illegal.

Hopefully Qt6's Multimedia module works better on Windows and macOS: https://doc.qt.io/qt-6/qtmultimedia-changes-qt6.html

**Comment by xiaoyifang** [comment]
> each failed pronunciation attempt increases GoldenDict's CPU usage

if this was introduced by ffmpeg ,I think it can be submitted to ffmpeg's official site.

**Comment by nonwill** [comment]
@xiaoyifang
Yes, you can go ahead and do this, first you will need to confirm that it is a ffmpeg issue, but failing that, you will get nowhere. Good luck to you🐷

**Comment by vedgy** [comment]
> cpu usage should be more easily identified compared to memory usage, I think take a dump of the application and check which
function takes the most of the stack trace should be enough to identify the cause.

I have attached GDB to the hanging-at-exit GoldenDict instance: there were hanging internal audio player backend threads. But the low-level FFmpeg backend code puts me off. I have implemented the Qt Multimedia backend precisely to avoid dealing with it. Since Qt Multimedia works almost perfectly for me, I am going to stick to it and let someone else fix the FFmpeg backend bug.

**Comment by vedgy** [comment]
> So it is a fact that ffmpeg 5 has some broken implementations which are not compatible with older versions.

Not so obvious. GoldenDict may have used the FFmpeg API in an unsupported way. It might have worked fine with older library versions by chance, then innocuous internal FFmpeg changes revealed the bug in GoldenDict's glue code.

**Comment by nonwill** [comment]
> It might have worked fine with older library versions by chance

ffmpeg 4 have the issue too. 
I am experiencing this issue two months ago with an mp3 file in eng-eng_LongmanPron(dsl format, wav files in .dsl.dz.files.zip) , the ffmpeg4 versions on three platforms are all bad. But all other files I tested in the same dictionary has no problem.
